### ISSUE TITLE
Hyper parameter search with real data set - a concern on CI values...

### ISSUE THREAD

**Issue by rmaanyam** [liupei101/TFDeepSurv#3]
Hi liupei101,

1. I tested the hyper parameter search piece with real data set - using [hyopt script](https://github.com/liupei101/TFDeepSurv/tree/master/bysopt) - as per instructions at [`bysopt`](https://github.com/liupei101/TFDeepSurv/tree/master/bysopt). It works fine and generates a list of hyper parameters in log file (for example: log_hpopt_xxx.json) - similar to the [one](https://github.com/liupei101/TFDeepSurv/blob/master/bysopt/log_hpopt.json) generated with simulated_data.

2) a) I loaded the real data - per [4.2.1](https://github.com/liupei101/TFDeepSurv#422-traning-or-testing-tfdeepsurv-model), [refer code snippet# 2  below] 

b) included the best parameters obtained from 1 (refer example parameters in code snippet# 1 below ]  to the real data set,  and initialized the neural n/w, as per [4.1.3](https://github.com/liupei101/TFDeepSurv#413-initialize-your-neural-network)

c) trained the n/w as per [4.1.4](https://github.com/liupei101/TFDeepSurv#414-train-neural-network-model) - via model.train and/or model.learn (tried both too), and attempted to evaluate performance and got learning curves - as described in [4.1.4a](https://github.com/liupei101/TFDeepSurv#414a-training-via-modeltrain) and [4.1.4b](https://github.com/liupei101/TFDeepSurv#414b-training-via-modellearn). Up to this point works great with no issues. 

However, I have a concern/question on finanl CI value mismatches - final ones on learning curve or from [4.1.5 ](https://github.com/liupei101/TFDeepSurv#415-evaluate-model-performance) - are lower than the ones generated in log_hpopt_xxx.json file. [For example,  Final CI values thru model.train or model.learn are way lesser than the ones mentioned in code snippet # 1 values;  ci_train and ci_valid are supposed to be in the range 0.7745, 0.7485 respectively - but I'm seeing in the range of 0.65 - 0.7 (Ci_train_set) and 0.55- 0.62 (ci_test_set). 
	 
**Question for you:** 
Why is it not able to generate the ci values - close to ones obtained in search-space in json file?  am i missing anything? One thing, I coudl see is - in [hyopt script](https://github.com/liupei101/TFDeepSurv/blob/master/bysopt/hpopt.py), the spilit seed = 42; but there is no seeding applied when I ran with real data, as per [4.2.1](https://github.com/liupei101/TFDeepSurv#421-load-real-world-data), using one time split (refer code  snippet# 2 below) - so, would that be a cause for total mismatch or lower ci values? Can you thow some light on this and point me to the right direction - if any thing missing...  

Hope I explained the issue clearly - if you need more info ro have any questions for me, please let me know. Appreciate your help/suggestions, pls.   
   
 
**Code snippet# 1:** 
`{"params": {"dropout": 1.0, "activation": "tanh", "L2_reg": 0.00027453142563396016, "learning_rate_decay": 1.0, "L1_reg": 0.0008515639202782554, "num_rounds": 1500, "learning_rate": 0.6, "optimizer": "adam"}, "ci_validation": 0.748552011474626, "ci_train": 0.7757347200677045},`  

**Code snippet# 2:** 
`train_X, train_y, test_X, test_y = load_data('full_data.csv', excluded_col=['ID'], surv_col={'e': 'event', 't': 'time'}, split_ratio=0.8)`

**Comment by liupei101** [comment]
I think it is related to some basics about model generalization.

Have to mention that the hyper-parameter tuning, in hyopt script,  is carried out on the training set, and each set of hyper-parameters is evaluated by the validation set which is randomly selected from the training set. 

Now, some questions come to us: should I trust the best parameters searched by the hyper-parameter tuning? or is the value of ci on a validation set enough to represent the goodness of a hyper-parameter set?

If the answer is "yes", then it is suggested that we must do k-fold cross-validation when searching each set of hyperparameters, which would need more time to calculate. It means that the result of hyper-parameters searching is more convincing and stable, not sometimes bad or sometimes good.

At last, it is better to test the difference of data distribution between your training set and test set. If the data distribution is close, then the performance on the test set will also be close to the `ci_validation` that you obtained with the best hyperparameters set.

**Comment by rmaanyam** [comment]
Thanks, sounds good.
I want to double-check with you to make sure I understand this part well:
Is the [simulated_data](https://github.com/liupei101/TFDeepSurv/blob/master/bysopt/simulated_data_train.csv) of [bysopt/hyopt](https://github.com/liupei101/TFDeepSurv/tree/master/bysopt) the same or part of train_data generated in [4.1](https://github.com/liupei101/TFDeepSurv#41-runing-with-simulated-data)?? And probably, the test_data of [4.1](https://github.com/liupei101/TFDeepSurv#411-import-packages-and-prepare-data) is of same kind of distribution as train_data, right? I guess, that may be the reason for getting very convincing ci-value matches with simulated_data example. Please correct me if I'm wrong...

- let me explore more on data distribution and/or splits or so and see if it helps.
